### ISSUE TITLE
[BUG] Clearing all data does not quite clear all data (revealed during export)

### ISSUE THREAD

**Issue by Gitoffthelawn** [darkreader/darkreader#4722]
### **Describe the bug**

In v4.9.26 with the v5 preview enabled, there is an option to clear all data.  It *almost* works, but when I then performed an export, there was 1 site still listed in the `siteList` section of the export.

### **Steps to reproduce**

- Enable the v5 preview mode in v4.9.26
- Use DR regularly. including enabling and disabling sites for it
- Click on the option to clear all settings
- Perform a settings export
- Open the export file

### **Expected behavior**

All data will be cleared, thus generating an export file with only default values.

### **Actual behavior**

The export file included 1 site that was previously excluded from being processed by DR.

### **System Information:**

- OS: Windows
- Browser: Firefox 84
- Dark Reader version: 4.9.26, with v5 preview enabled

**Comment by Gusted** [comment]
Hi @Gitoffthelawn!

Did you do anything special, we just set the settings to the `default` ones without any special processing. I tried it myself and all data(except previewnewdesign) where the default ones including the sitelist.

**Comment by Gitoffthelawn** [comment]
@Gusted Yes, this is a tricky bug to report because I don't have sufficient data to fully define the second step in the report.

I used DR as I normally would, and then discovered the issue.

This might help: Even after clearing all DR settings, going to the 'Site list' page in the v5 preview UI showed the non-deleted website domain.  So the DR export file correctly matched DR's internal data.  IOW, I don't think it's an export bug... I think it's a bug in the 'Reset settings' routine.

**Comment by Gusted** [comment]
> This might help: Even after clearing all DR settings, going to the 'Site list' page in the v5 preview UI showed the non-deleted website domain. So the DR export file correctly matched DR's internal data. IOW, I don't think it's an export bug... I think it's a bug in the 'Reset settings' routine.

Problem: It's 1 line of code, that reset it's to `DEFAULT_SETTINGS`
https://github.com/darkreader/darkreader/blob/master/src/ui/popup/manage-settings-page/reset-settings-button.tsx#L23
And `DEFAULT_SETTINGS` include the `siteList: []`. I'm unsure where in the process this bug could occur.

**Comment by Gitoffthelawn** [comment]
Well clearly it needs two lines of code!  😃 😃 😃 

All kidding aside, let's think about this together...

Here are 3 initial thoughts:

1. It's acting like data is stored in array, but instead of actually removing all the contents of the array, it is setting the array size to 1.  Thus the first item still appears.  That's how it's behaving, but I still need to go through the code to see what's happening.
2. Another possibility, although it seems unlikely, is that the domain name was stored in a variable containing the 'Site list' ***input field*** (not the 'Site list' array).  I think that's a long shot, but worth mentioning the possibility.
3. A browser bug!

I'm going to take a look at what is called by `props.actions.changeSettings` when I have a little more time.

**Comment by Gusted** [comment]
> I'm going to take a look at what is called by `props.actions.changeSettings` when I have a little more time.

https://github.com/darkreader/darkreader/blob/master/src/background/extension.ts#L263

>     1. It's acting like data is stored in array, but instead of actually removing all the contents of the array, it is setting the array size to 1.  Thus the first item still appears.  That's how it's behaving, but I still need to go through the code to see what's happening.

That indicates to be a browser bug as array should override each other.



>     2\. Another possibility, although it seems unlikely, is that the domain name was stored in a variable containing the 'Site list' _**input field**_ (not the 'Site list' array).  I think that's a long shot, but worth mentioning the possibility.

I'm not aware that this is possible.



>     3\. A browser bug!

Possible, but unsure where as I need a reproduce-able case.

**Comment by Gitoffthelawn** [comment]
@Gusted After more testing and experimentation, there is something seriously amiss with DR data storage.  At this point, I can replicate the bigger issue 100%, but I have not determined what is causing the larger problem.  It's most likely the browser or DR.

Would you prefer I create a separate issue, or continue our discussion here?  It's possible the other issue I discovered (which is much larger) may be causing this issue, but it's also possible it's a separate issue.

**Comment by Gusted** [comment]
> @Gusted After more testing and experimentation, there is something seriously amiss with DR data storage. At this point, I can replicate the bigger issue 100%, but I have not determined what is causing the larger problem. It's most likely the browser or DR.
> 
> Would you prefer I create a separate issue, or continue our discussion here? It's possible the other issue I discovered (which is much larger) may be causing this issue, but it's also possible it's a separate issue.

I think it is appropriate to continue the discussion here as this issue doesn't really have a purpose other than 'Oh no not all data is gone' which can be extended with your discussion.

**Comment by Gitoffthelawn** [comment]
@Gusted Sounds good.  Give me a few minutes to write something cohesive.  I just finished creating STR that, on my end, reproduces the issue 100%.  I'll post the STR before testing on a browser with default settings so I can get your feedback before I do that.

**Comment by Gitoffthelawn** [comment]
Here are my STR for the larger issue.  Note that as I just discovered all this, I have not performed testing on Firefox with default settings yet.

Setup:

- Browser: Firefox v84.0.2 with sync disabled
- Dark Reader: v4.9.26 with v5 preview enabled

STR (note: this is destructive to current DR settings!):

1. Reset DR settings via the following sequence in DR's v5 preview popup: Settings > Manage settings > Reset settings
2. Import the DR settings file attached to this post (you can rename it to have a `.json` extension; GitHub does not currently allow `.json` attachments... lol)
3. In a new tab, go to https://example.com
4. Open the DR popup

Issues:

A. Assuming the syntax is all correct in the file I provided for step 2 *(please verify!)*, I would expect the DR popup to indicate that https://example.com is handled by the DR theme named `Custom: Static 100/100`.  Instead DR shows `Theme for all websites`.  Let's call this 'Issue A', and increment issue letters from here.

B. When clicking the DR button with the `Theme for all websites` label, a dropdown appears.  One of the themes is called `Theme 1`, and yet all themes are assigned names within the file provided.

C. When clicking the DR button with the `Theme for all websites` label, a dropdown appears, but it does not show all the themes defined in the file provided.

D. This is the big issue to which I previously referred: When clicking the DR button with the `Theme for all websites` label, a dropdown appears, but it contains theme names from previous configurations of DR!  Keep in mind we thought we cleared all DR data in step 1!

I tested all of the above 5 times, and reproducibility was 100% each time.


[Dark-Reader-Settings - final test for 4 custom themes.txt](https://github.com/darkreader/darkreader/files/5793421/Dark-Reader-Settings.-.final.test.for.4.custom.themes.txt)

**Comment by Gitoffthelawn** [comment]
@Gusted Just posted STR.  Do you get pinged if I don't include @Gusted in posts?  If not, here's your ping! :)

**Comment by Gusted** [comment]
> @Gusted Just posted STR. Do you get pinged if I don't include @Gusted in posts? If not, here's your ping! :)

Don't worry I've made sure to get every comment on this repo, including smartwatch/phone/desktop/laptop email notifications, it can be quite spammy sometimes :P.
![image](https://user-images.githubusercontent.com/25481501/104137483-5bcc0400-539d-11eb-9781-27a207e79974.png)

**Comment by Gitoffthelawn** [comment]
Smartwatch notifications!  LOL!   My watch has a little dial I turn to wind it up!

**Comment by Gusted** [comment]
>     2\. Import the DR settings file attached to this post (you can rename it to have a `.json` extension; GitHub does not currently allow `.json` attachments... lol)

Before continuing because I'm reproducing the issue as well, you imported the settings trough the hack mentioned over at #3927 so it will actually import the settings.



> Smartwatch notifications! LOL!

Well yeah, I don't always have my phone/laptop/desktop on me, but with my Smartwatch it's a different story :P

**Comment by Gitoffthelawn** [comment]
> Before continuing because I'm reproducing the issue as well, you imported the settings trough the hack mentioned over at #3927 so it will actually import the settings.

Yes, specifically I used this hack that I provided:
https://github.com/darkreader/darkreader/issues/3927#issuecomment-716642446

**Comment by Gusted** [comment]
> A. Assuming the syntax is all correct in the file I provided for step 2 _(please verify!)_, I would expect the DR popup to indicate that https://example.com is handled by the DR theme named `Custom: Static 100/100`. Instead DR shows `Theme for all websites`. Let's call this 'Issue A', and increment issue letters from here.

I don't get this issue, it will just show it's `Custom: Static 100/100`, nor the other issues. I only got it their was some data inconsistency between some functions, but it was already reset. I might suspect the browser is not saving the data properly, you can confirm this by going to `ctrl + shift + a` -> Extensions -> Setting icon -> Debug Add-ons -> Click on Inspect by Dark Reader.
Now type something in the console and enter it to mark that this is where you started to repeat the steps, after importing it should log an `saved` message alongside the new settings, does this include the 4 presets?

Regards,
Gusted

**Comment by Gitoffthelawn** [comment]
Interesting... hmmm.... can you edit the second sentence you wrote... I'm having difficulty understanding what you meant.

I'll perform the other steps you requested right now.

**Comment by Gitoffthelawn** [comment]
Steps completed... log forthcoming.  Need to format it to make it readable for you.

**Comment by Gitoffthelawn** [comment]
Log attached for you.  

The log includes using DR to clear all its own data and then performing the import, and then waiting 5 minutes.  I'm seeing some other possible issues in the log, but my first scan of the log indicates nothing regarding the above issues.  I need to review the log a more thoroughly though.

After performing the actions to generate the log, the problematic results I previously reported are all consistent.

[darkreader_log_001_for_Gusted.txt](https://github.com/darkreader/darkreader/files/5793477/darkreader_log_001_for_Gusted.txt)

**Comment by Gitoffthelawn** [comment]
Don't attempt viewing the log on your smartwatch! :) :) :)

**Comment by Gusted** [comment]
> Log attached for you.
> 
> The log includes using DR to clear all its own data, performing the import, and then waiting 5 minutes.
> 
> I'm seeing some other possible issues in the log, but my first scan of the log indicates nothing regarding the above issues. I need to review the log a more thoroughly though.
> 
> After performing the actions to generate the log, the problematic results I previously reported are all consistent.
> 
> [darkreader_log_001_for_Gusted.txt](https://github.com/darkreader/darkreader/files/5793477/darkreader_log_001_for_Gusted.txt)

Log seems fine...... other then known errors. Presets are being loaded correctly... However those errors might indicate that new data is not send to UI/Tabs. Could you try on a new fresh profile?

**Comment by Gitoffthelawn** [comment]
Do you know of any good tools to inspect sqlite files created by Firefox?

I tried "Database Browser by ETL-tools" and "sqliteman", and both choked on Firefox sqlite files.

I then tried "DB Browser for SQLite" and it could handle the files a little better.  I wrote SQL code to gather the data from the appropriate table within DR's sqlite database, but the data is stored in binary blobs, which seem to be compressed or stored in some weird encoding.  It's almost ASCII, but not quite.

Not one to give up, I exported the relevant blobs as binaries and pulled them up in a hex editor where I could easily twiddle the bits.  I confirmed, with 100% certainty, that the DR sqlite database store by Firefox has old data in it.  There are 2 possibilities:

1. A privacy and/or security issue in DR and/or Firefox, but not a functional issue if the data is remnant data that is not actually being used.
2. A privacy and/or security issue in DR and/or Firefox, and a functional issue if the data is actually being used.

Given the tests results from our discussions earlier in the thread, I think I've ruled out possibility number 1.  The data clearly gets used.  It also explains why you are following the same steps, but not seeing the issues right now: because your DR sqlite database would be different from mine (although, if everything was working correctly, after using DR to clear its own data, our databases *should* be almost identical - only any embedded UUIDs and timestamps would vary).

Personally, I should be able to workaround the issue by uninstalling DR, deleting the DR sqlite database (if it persists), and then reinstalling DR.  Of course, that doesn't actually fix any bugs.  Doing so also does not address any other issues that may be revealed by the log I uploaded.

**Comment by Gusted** [comment]
> Do you know of any good tools to inspect sqlite files created by Firefox?

When I need some data I use command line tools to search for it. No external tools, just `cat` and old good `grep`.

> Not one to give up, I exported the relevant blobs as binaries and pulled them up in a hex editor where I could easily twiddle the bits. I confirmed, with 100% certainty, that the DR sqlite database store by Firefox has old data in it. There are 2 possibilities:

Aha...



> Given the tests results from our discussions earlier in the thread, I think I've ruled out possibility number 1. The data clearly gets used. It also explains why you are following the same steps, but not seeing the issues right now: because your DR sqlite database would be different from mine (although, if everything was working correctly, after using DR to clear its own data, our databases _should_ be almost identical - only any embedded UUIDs and timestamps would vary).

Theoretical and my current knowledge about their storage it should.



> Personally, I should be able to workaround the issue by uninstalling DR, deleting the DR sqlite database (if it persists), and then reinstalling DR. Of course, that doesn't actually fix any bugs. Doing so also does not address any other issues that may be revealed by the log I uploaded.

Not sure if the bug like you said will be gone.

**Comment by Gitoffthelawn** [comment]
> When I need some data I use command line tools to search for it. No external tools, just `cat` and old good `grep`.

Yes, those would normally work, but the data is not being stored in an easily readable encoding like ASCII.  I had to do a fair amount of bit twiddling to decode it completely.  How the bits looked, it was like the data was compressed with a very inefficient compression algorithm.

> Aha...

👍🏾 

> Theoretical and my current knowledge about their storage it should.

Thanks.  Good to know. 

> Not sure if the bug like you said will be gone.

I'm not sure either, but I'm going to give it a try.  I think it will resolve it (temporarily) because the old data has to be coming from *somewhere*, and the only place I've confirmed it does exist is within DR's sqlite database file.  Of course, I've only spent the time to decode all the binary blobs in that one file, and not within every file within the profile.  That said, the techniques I used to isolate where to look revealed no other hits.

**Comment by Gitoffthelawn** [comment]
I uninstalled DR which caused DR's existing sqlite database to be deleted by Firefox.  I then reinstalled DR, and performed the steps outlined earlier.  With a fresh DR sqlite database, everything worked as expected.

I think there are some significant issues, but tracking them down will likely not be a trivial task.

Did you have a chance to read the log I uploaded?  Any thoughts on that?

**Comment by Gusted** [comment]
> Did you have a chance to read the log I uploaded? Any thoughts on that?

https://github.com/darkreader/darkreader/issues/4722#issuecomment-757570348
